Addiction Rework, OD Changes, Reagent Tweaks and Additions

FalseIncarnate opened the pull request:
Completely reworks how Addictions are handled
- Addictions now are tracked based on the amount of reagent taken over
time
- Addictions decay very slowly on their own if none of the reagent is present in your system
 - This means taking multiple small doses will cause addiction if not paced properly
- Addiction effects now are relative to the total amount of the addicted reagent that has been taken
 - The addiction threshold for a reagent defines the increments between addiction stages (Stage 4 = 4 x Threshold)
 - For example, 30 units of meth will cause a Stage 3 addiction, while 29.9 will only cause a Stage 2 addiction
- Addiction effects now begin with the highest number stage and decrease until the addiction is broken
 - This means you would start at Stage 4 and work down to Stage 1 if the conditions are met

Adjusts how Overdosing works
- Overdose effects now stop if you have less than the overdose threshold for the given reagent in your system
 - Previously, overdose effects would continue until all of the reagent was removed from the body
- After dropping below the overdose threshold, reagents will process as normal as long as they do not surpass the threshold again.

Tweaked some Synthetic-oriented reagents
- Synthetic drugs no longer will cause Toxins damage
 - The toxins damage for these has been replaced with either another form of damage or side-effect
- Degreaser no longer cures Toxins Damage in Synthetics
 - Degreaser still acts as a chemical purgative and lube cleaner, but
helps cure certain side-effects much like Coffee does for organics.

Sterlizine recipe changed to require antihol and chlorine
- Effectively the same as the old recipe, except actually mixable
- Fixes #2030

New Reagent: Fixer!
- Temporarily prevents addiction effects while in the system
 - Yes, it is a Fallout chem
- Can be processed by both Synthetics and Organics
- Can be overdosed, but is not addictive
- Recipe: Doctor's Delight + Morphine + Sugar

New Reagent: Cold Turkey!
- Decays addictions faster, even if the addicted reagent is in your system, but has a plethora of side-effects
 - Side effects include: Brain Damage, Vomiting, Loss of Stamina, Confusion, Toxins buildup, Shortness of Breath, Flatulence, Jittering, Blurry Vision, Itching, Uncontrollable Screaming,  Temporary Deafness, Hunger, and Increased Body Temperature
 - Side-effects only occur if there are any addictions and are selected randomly, otherwise this reagent does nothing
- Can be processed by both Synthetics and Organics
- Cannot be overdosed and is not addictive
- Recipe: Nutriment + Haloperidol + Ether + Cryostylane (Catalyst)
 - Must be chilled to below 273K to react

Increases max_damage of MMI Holders to 200 (from 60)
- Now matches the max_damage for brains
 - Should make IPC brain damage more consistent with organic brain damage

Gives IPCs the NO_POISON flag
- This should make IPCs immune to toxins damage again. Fixes https://github.com/ParadiseSS13/Paradise/issues/2151.

Fox-McCloud commented on code/modules/reagents/newchem/drugs.dm:
```diff
@@ -551,7 +551,7 @@
 	M.Jitter(20)
 	M.Dizzy(20)
-	M.adjustToxLoss(5)
+	M.adjustBrainLoss(2)
```
Can IPCs even *get* brain damage?

FalseIncarnate replied:
Yes they can, Posibrains function as a brain organ for them. Brain damage can be repaired via surgery or via the administering of Liquid Solder.

IPC brain damage was actually quite scary before the IPC update, since it wasn't possible to revive an IPC who died of brain damage. Mark has assured me that it is fixable via surgery though, so that shouldn't be a concern anymore.

Fox-McCloud replied:
Just checked; they have the same problem brains did; their damage caps at 60.

You'll have to set the MMI holder's max health to 200, same as brains, so it has the same impacts/effects.

Also, you'll probably want to bump up the brain damage here by 25%---2 brainloss will not be 2 brainloss; it'll be 1.8 because all robotic organs have 20% damage resist; 2.5 brainloss will effectively deal 2 brain damage per cycle.

Markolie replied:
Very nice. Thank you for taking care of this! If you want, you can also add NO_POISON to the IPC flags. And yes, increasing the health of the mmi_holder would be wise.

Fox-McCloud commented:
:+1: to the IPC drug fixes
:+1: to the OD changes
:+1:  to sterilizine fix

:-1: to the addiction rework. 

I do generally think that our current addiction system isn't *ideal*, but this is a step in the wrong direction. There's a ton of chems that are utilized by traitors and the likes that are addictive; these are all specifically balanced around the addiction threshold either being circumventable or not trigger with the current amount. This system makes it literally completely unavoidable in a lot of scenarios and HEAVILY nerfs combat related drugs (which don't need nerfing, IMO)---stuns are still king in SS13, and there should be an effective way to mitigate them; this heavily neuters a lot of them, as it's now pretty much impossible to use them without suffering some side effect.

I disliked this system, quite heavily, when you first mentioned it and when you detailed exactly how it works, and I dislike it still now.

The purpose of addictions and ODing is to prevent long-term extended use of chems without pacing yourself; this makes it impossible to do so as you're always going to get walloped upside the head, regardless of whether or not you pace yourself (unless the pacing is to a ridiculous degree)---combat chems are hardly ever/effectively utilized or abused; they don't need a nerf like *this* now.

Fox-McCloud commented on code/game/atoms.dm:
```diff
@@ -464,4 +464,7 @@ its easier to just keep the beam vertical.
\ No newline at end of file
+		O.show_message("<span class='game say'><span class='name'>[src]</span> [atom_say_verb], \"[message]\"</span>",2)
+
+/atom/proc/update_addictions(var/datum/reagent/R, var/amount)
```
No offense, but uh..why in the world did you make a mob related proc---a proc for every single atom?

Aside from my disagreement with the addiction system, this should be under carbon, not every single atom.

FalseIncarnate replied:
There was initially a reason, but I believe I moved some things around partway and forgot to adjust this. Will fix this evening

Fox-McCloud commented on code/modules/reagents/newchem/medicine.dm:
```diff
@@ -1040,3 +1045,103 @@ datum/reagent/haloperidol/on_mob_life(var/mob/living/M as mob)
+	color = "#00DD33"
+	process_flags = ORGANIC | SYNTHETIC
+	metabolization_rate = 0.3
+	addiction_threshold = 0		//can't become addicted to an addiction treatment
```
this is 0 by default; no point in defining it as 0 again here.

Regen1 commented:
>stuns are still king in SS13, and there should be an effective way to mitigate them; this heavily neuters a lot of them, as it's now pretty much impossible to use them without suffering some side effect.

Perhaps we should rework that part as well then? Perhaps, you know.. remove instant stuns from.. tasers...

:+1: For the addiction system, if this nerfs combat chems THAT much for traitors (which I rarely seem get utilized by traitors, by the way) we can play more around with the numbers for those chems, since they are traitor only.
The addiction thresholds for traitor chems are not set in stone, and can be tweaked later

Fox-McCloud commented:
@Regen1 

Not every combat chem is trator only--likewise, it's not only tasers that stun--there's batons, lube, tabling, pushing, vampire gaze--the list goes on and on and on. It's not a fact of the numbers need to be played with or not--that's already in a decently good spot--it's the fact that combat chem's usage generally doesn't need to be negated, at the moment.

About the only exceptions is Virology--but that's a problem with virology, not the entire stun system, chem addictions, etc.

FalseIncarnate commented:
The system uses global vars for the decay rate, so tweaking it should be easy if we can agree on a value. Currently, it decays by 0.1, which does mean addictions last a lot longer. I could bump this to 0.5 or something, if that would be agreeable.

As for traitors, I had an idea of perhaps giving them a traitor-only chem that completely wipes all addictions so they can use addictive combat drugs without suffering withdrawal. This would be the least snowflake code method of handling that concern that I can think of.

Meth virus shouldn't prompt addictions if it uses add_reagent, but I will need to double check that.

That said, the intent is not to necessarily balance the combat drugs but rather make addictions function more logically and be harder to avoid than repeatedly taking 0.1 units under the threshold

Fox-McCloud commented:
While things functioning more logically is a generally a good thing---the whole purpose of stun reducing drugs is to enhance combat effectiveness, to that end, addictions OD, and the chem's inherent downsides generally punish their usage over the long term. I don't believe we need a system whereby this because something else the user has to manage/be wary of, constantly.

I can understand the want to rework the current addiction system, but this isn't a good way to go about it, IMO.

Iamgoofball commented:
This is terrible, you just defanged addictions and ODing completely. Get good and deal with the ODs and addictions instead of crying for a nerf when you stub your toe because you can't do every drug at once.

FalseIncarnate commented:
First off, this doesn't really "defang" addictions, though it does arguably do so for overdosing. In fact, currently it actually makes addictions WORSE because it processes them many times more per stage.

Second, nerfing the addiction system was not the intent, and largely is not present. The only real "nerf" is that you don't suffer every stage of the addiction if you take 11 units of meth (1 above the threshold), but rather only the first stage. Additionally, repeatedly taking 9.9 units of meth (0.1 below the threshold) will trigger addictions, meaning this actually buffs addictions by countering the powergaming aspect of remaining JUST below the threshold.

Third, this does absolutely nothing to the ability to take and become addicted to multiple drugs at once. I'm not sure where that idea came from, but it is outright false.

Fourth, this doesn't affect how ODs and addictions are dealt with, save for the addition of new chems designed to help doctors (or less savory types) treat addiction effects.

Finally, this PR was not created in response to any calls for a nerf or desire for hug-boxing the system. It was made based on an idea and some feedback on said idea in order to make the system more dynamic and interesting from both a gameplay and roleplay perspective.

If you wish to provide any sort of meaningful feedback, such as suggestions of how to improve this, it is entirely welcome. Otherwise, please refrain from comments that effectively total to "This is shit, you are shit" like your last comment did, Goofball, as they do little to stimulate the discussion of the proposed changes.

Iamgoofball commented:
Okay.

Why did you nerf the shit out of overdosing then?

FalseIncarnate commented:
To be honest, it was largely at Fox's urging, though I do admit that I had already been thinking that overdoses should stop once you drop below the reagent's overdose threshold.

I'm not sure how realistic either way is, but it seemed logical that an overdose should only occur while you have more than the safe dosage in your system. Once it drops below the threshold, your body should be able to metabolize the remaining amount normally, since it is no longer in excess. This also means a very small overdose (say, by 0.1 units) is punished less severely than an excessive overdose. While doctors should be careful about how much of a medicine they give a patient, this should hopefully mean they waste less chemicals if they misjudge their timing by a cycle or two.

Again, not really sure how realistic it is (I don't study medical / pharmaceutical chemistry), so it was a change made to make things more sensible. While this does largely cut down on the danger of overdoses, we can easily bump up the overdose effects to compensate or undo this particular change in the future should it turn out to be too weak in actual gameplay.

FalseIncarnate commented:
Updated (updating original post too) with a new commit:

Removes the update_addictions proc from /atom level
- Now located at /mob/living/carbon level only
- Adjusts the code that calls update_addictions to match

Increases max_damage of MMI Holders to 200 (from 60)
- Now matches the max_damage for brains
 - Should make IPC brain damage more consistent with organic brain damage

Gives IPCs the NO_POISON flag
- This should make IPCs immune to toxins damage again

Adjusts Addiction Decay rates
- ADDICT_DECAY: 0.1 -> 0.5
 - The previous value made addictions last TOO long
- ADDICT_DECAY_CT: 0.9 -> 1.0

Removed redundant threshold defines from Fixer and Cold Turkey

Also confirmed that the meth virus symptom does not cause or worsen addictions due to using add_reagent. The presence of meth in your system will still stave off the addiction effects as normal though.

Markolie commented on code/modules/mob/living/carbon/alien/life.dm:
```diff
@@ -22,6 +22,49 @@
+	reagents.addiction_tick++
+	updatehealth()
+
+	return
```
Unnecessary return.

Markolie commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -686,4 +686,37 @@ var/list/ventcrawl_machinery = list(/obj/machinery/atmospherics/unary/vent_pump,
+				wear_mask)
+
+/mob/living/carbon/proc/update_addictions(var/R, var/amount = 0)
+	if(isbrain(src) || isslime(src))		//Only human and xeno types can handle addictions
```
This proc should probably be in the life.dm file, and I'm not too sure about having this check hardcoded here. Wouldn't a variable on the mob make more sense?

Markolie commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -850,6 +851,49 @@ var/global/list/brutefireloss_overlays = list("1" = image("icon" = 'icons/mob/sc
 
 	return //TODO: DEFERRED
 
+/mob/living/carbon/human/proc/handle_addictions()
```
There seems to be a lot of copypasta going on between all the Life() files and in this handle_addictions() proc. I think it's better to move these procs to a /mob/living level, and then determine what should(n't) run based on mob variables.

FalseIncarnate commented:
I'll fix the merge conflicts after work (already running a bit late)

SamCroswell commented:
The second commit should probably be a separate PR.
Keeping it atomic 'n such.

FalseIncarnate commented:
@SamCroswell That probably should be, yes (at least the parts that don't tweak the new stuff), but was done at the urging of a maintainer since it related to the previous commit's changes to some synth reagents no longer causing/healing toxins damage as IPCs weren't meant to be taking toxins at all.

The PR that added those was more or less building on an oversight/bug and touting it as intended functionality, so the change to the IPC species flag could be viewed as fixing that bug. The MMI thing was mostly to bring synthetic brain damage values in line with organics to maintain a balance.

Fox-McCloud commented:
I'll be testing this out, locally, to get a better feel for it, soon before I formulate my final opinion on it.

TheDZD commented:
Played around with this a little bit, and I honestly don't have a very conclusive opinion on it. Addictions not magically being based off of reaching a specific threshold of reagents inside of a humanoid makes sense, although the main problem I can see is that there might be issues for traitors items that use chemicals that would previously not addict you to them.

FalseIncarnate commented:
Any items that previously used chemicals and did not cause an addiction would function roughly the same, except in cases of repeated use, since no values were changed for the thresholds or the contents of the item.

Further, it would be trivial to add a traitor-only chemical which completely clears any and all addictions when ingested should there be concerns regarding traitor items and addictions. Traitor items could be tweaked to have an incredibly small amount of this chemical (though this would effectively make the original chem last no more than a single cycle shorter than previously), which would clear the addiction list when processed. Addictions are only added/increased when the reagent is first transferred to the body, so this chemical would only need to process a single time to undo the addiction update (ie, no more than 0.1 units would be needed).

Fox-McCloud commented:
@FalseIncarnate 

Alright, I reviewed and tested this....while it's an interesting system, I can't say I'm personally a fan of it, to be honest. IMO, it doesn't address the concerns of why addictions exist in the first place---furthermore, I feel it makes the downsides of utilizing combat chems too steep. No matter which system we have, ultimately, addictions being circumventable will be a thing (and I don't find that inherently problematic, TBH).

I'm sorry, I know you put a lot of thought and effort into this, but I really can't say I support the addiction system you've outlined here.

I really like all the other changes though.

FalseIncarnate commented:
@Fox-McCloud 
The intent of the rework was not to address the concerns of the existence of addiction mechanics, but rather to address the way they were handled to make them a more dynamic experience. Before, addictions were simply a soft overdose threshold: go above X units in your system and you suffer all 4 stages regardless of how little or much you went over. This meant someone who had 10.1 units of meth (a somewhat common combat drug) in their system would suffer through all 4 addiction stages which progressively got worse as they went through each. Under this rework, someone who only took 10.1 units would only need to sit through a (rather short) first stage which is actually quite tame, which in some regards would encourage them to use the combat drugs since it is less punishing for them to go over the threshold by such a small amount.

The only people who truly feel the pain with the new system are the people who are consistently using the chem. A traitor who takes a dose or two to give them the edge over validhunters/security is going to be able to ignore the harsher addiction stages under the rework unless they are taking doses that border on the overdose threshold anyways simply by way of numbers. Meanwhile, Druggy McChemist is going to be face down in maintenance since he's been taking the drug all shift long.

The system was never intended as a balance change, though given the nature of the beast it is difficult to change much without affecting balance in some regard. That said, I know you have said you don't care for or support the rework as is, and can respect that. If you were able to communicate the portions of it that you disagree with, or a system that would be more agreeable, I'd be willing to work with you to make any changes to the rework that would improve the system all-around.

fludd12 commented:
Personally, I am 100% for it. I know of perhaps a single chemical that I've ever seen used for a low dosage overdose attack, and that is omnizine. Everything else people go for upwards of fifty units. Also, I support any system that makes addiction less of a >Whoops, looks like I have to chug triplecitrus for the next hour.< And more of a >I've taken nothing but ultralube this entire shift and now my posibrain is in my left foot<

Iamgoofball commented:
you realize you can also prevent addiction effects by taking the drug itself right

FalseIncarnate commented:
Taking the reagent will still delay the effects, except it can also worsen the addiction once it kicks back in with this rework.

Iamgoofball commented:
no, not that drug

i'm talking about what you're addicted to

you can sate your addiction while it's in your system

FalseIncarnate commented:
That's what I said.

Fixer doesn't delay, it just skips the effect entirely while the addiction degrades.

Taking the addicted drug pushes back the addiction effects and freezes the decay, but also increases the addiction tracker since you took the drug.

Iamgoofball commented:
Uh, what?

If you're using my addiction code, it should set you back to 0 addiction on that drug if you take whatever you're addicted to again and never increment it until it's fully out of your system.

FalseIncarnate commented:
First off, I don't think we were using your addiction code before. Previously, taking a drug set the addiction back by setting it to -15, which would not change until the reagent was fully metabolized/removed. Then it would begin incrementing normally, though that -15 gave a buffer zone before the effects kicked in again. They would increase until about 40, at which the addiction would break and be cleared. While increasing, it would start at stage 1 and then go through all 4 stages in ascending order.

The new system does not give the -15 buffer zone when taking an addicted reagent, instead starting the effects back up as soon as the reagent is completely gone from the body (within about 2 seconds of being cleared usually). It works through the stages in DESCENDING order, starting at whichever stage is appropriate based on which multiple of the addiction threshold that drug's counter exceeds (ie, start at stage 3 if above 30 units have been taken of a drug with a threshold of 10). Taking more of the drug while addicted increases the counter (potentially pushing your addiction to the next stage) while preventing it from decaying naturally and preventing addiction effects while it is in the body.

The fixer reagent simply prevents addiction effects while in the body, but is easily overdosed. It won't speed up the decay, so it's meant to aid medical in treating the higher addiction stages by buying them time while treating the damages. Cold turkey decays the addiction faster (like using calomel to clear out reagents) but at the cost of a wide variety of side effects which can become deadly, and does not prevent the addiction effects while active (so you can still rack up damages from the addiction plus the cold turkey side effects).

Addiction isn't useless with these changes and additions, but rather is a more dynamic and interactive mechanic for both the sufferer and medical team. Careful management can minimize the effects, but to completely negate the effects will require precise timing and dosages that will challenge powergamers.

The system is not useless, especially with the new Chems. The Chems are meant to give players options on how to handle the new system. Cold turkey isn't a smart choice to use in combat, as it can potentially stun you or cause more damages. Fixer is only really useful if you are suffering the worst stages of addiction, but you can't rely on it because of how quickly it metabolizes and the threat of an overdose on it.

Fox-McCloud commented:
@FalseIncarnate 

He's the one that implemented it, we have his addiction system--if it's not precisely the way it is on TG, it's because further amendments were made to it after @Iamgoofball ported this over for us---I know for  a fact that I never once touched any of the addiction mechanics except for the teleporty nature of Meth, and that's literally it.

I understand how your system works, @FalseIncarnate, I've read through it and play-tested it--I know you didn't explicitly intend for it to directly be a balance change, but inevitably system changes do become balance changes, and this is no exception---that doesn't change my opinion of it, in this case or instance, I'm afraid.

FalseIncarnate commented:
@Fox-McCloud

I was not attempting to repersuade you of this, nor attempting to explain to you in hopes you would change your mind. I was explaining it to @Iamgoofball to try and clear what came across as confusion from his end.

That said, the way that I explained the old system in my previous comment is correct for our code, as is visible in the newchem_procs.dm file at line 58 in the current live code. I don't know if that was how it was pre-goonchem or not, but I wanted to make sure we both were discussing it from the same point of reference.

FalseIncarnate commented:
Closing this, will break off the overdose and other changes from the addiction rework and re-submit those. Will try reworking addictions again in the future to hopefully address any issues people have with this attempt.
